group-coordinator/src/test/java/org/apache/kafka/coordinator/group/consumer/CurrentAssignmentBuilderTest.java

```diff
@@ -0,0 +1,558 @@
+        ), updatedMember.partitionsPendingRevocation());
+        assertEquals(mkAssignment(
+            mkTopicAssignment(topicId1, 4, 5),
+            mkTopicAssignment(topicId2, 7, 8)
+        ), updatedMember.partitionsPendingAssignment());
+    }
+
+    @Test
+    public void testTransitionFromNewTargetToAssigning() {
+        Uuid topicId1 = Uuid.randomUuid();
+        Uuid topicId2 = Uuid.randomUuid();
+
+        ConsumerGroupMember member = new ConsumerGroupMember.Builder("member")
+            .setMemberEpoch(10)
+            .setPreviousMemberEpoch(10)
+            .setNextMemberEpoch(10)
+            .setAssignedPartitions(mkAssignment(
+                mkTopicAssignment(topicId1, 1, 2, 3),
+                mkTopicAssignment(topicId2, 4, 5, 6)))
+            .build();
+
+        assertEquals(ConsumerGroupMember.MemberState.STABLE, member.state());
+
+        Assignment targetAssignment = new Assignment(mkAssignment(
+            mkTopicAssignment(topicId1, 1, 2, 3, 4, 5),
+            mkTopicAssignment(topicId2, 4, 5, 6, 7, 8)
+        ));
+
+        ConsumerGroupMember updatedMember = new CurrentAssignmentBuilder(member)
+            .withTargetAssignment(11, targetAssignment)
+            .withCurrentPartitionEpoch((topicId, partitionId) -> 10)
+            .build();
+
+        assertEquals(ConsumerGroupMember.MemberState.ASSIGNING, updatedMember.state());
+        assertEquals(10, updatedMember.previousMemberEpoch());
+        assertEquals(11, updatedMember.memberEpoch());
+        assertEquals(11, updatedMember.nextMemberEpoch());
+        assertEquals(mkAssignment(
+            mkTopicAssignment(topicId1, 1, 2, 3),
+            mkTopicAssignment(topicId2, 4, 5, 6)
+        ), updatedMember.assignedPartitions());
+        assertEquals(Collections.emptyMap(), updatedMember.partitionsPendingRevocation());
+        assertEquals(mkAssignment(
+            mkTopicAssignment(topicId1, 4, 5),
+            mkTopicAssignment(topicId2, 7, 8)
+        ), updatedMember.partitionsPendingAssignment());
+    }
+
+    @Test
+    public void testTransitionFromNewTargetToStable() {
+        Uuid topicId1 = Uuid.randomUuid();
+        Uuid topicId2 = Uuid.randomUuid();
+
+        ConsumerGroupMember member = new ConsumerGroupMember.Builder("member")
+            .setMemberEpoch(10)
+            .setPreviousMemberEpoch(10)
+            .setNextMemberEpoch(10)
+            .setAssignedPartitions(mkAssignment(
+                mkTopicAssignment(topicId1, 1, 2, 3),
+                mkTopicAssignment(topicId2, 4, 5, 6)))
+            .build();
+
+        assertEquals(ConsumerGroupMember.MemberState.STABLE, member.state());
+
+        Assignment targetAssignment = new Assignment(mkAssignment(
+            mkTopicAssignment(topicId1, 1, 2, 3),
+            mkTopicAssignment(topicId2, 4, 5, 6)
+        ));
+
+        ConsumerGroupMember updatedMember = new CurrentAssignmentBuilder(member)
+            .withTargetAssignment(10, targetAssignment)
```

Comment:
should this be with epoch 11 so that we recompute from initial state?

Reply:
you're right. good catch!